// @india-boundary-corrector/openlayers-layer/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.js
var index_exports = {};
__export(index_exports, {
  IndiaBoundaryCorrectedTileLayer: () => IndiaBoundaryCorrectedTileLayer,
  LayerConfig: () => import_layer_configs2.LayerConfig,
  getPmtilesUrl: () => import_data2.getPmtilesUrl,
  indiaBoundaryCorrectedTileLayer: () => indiaBoundaryCorrectedTileLayer,
  layerConfigs: () => import_layer_configs2.layerConfigs
});
module.exports = __toCommonJS(index_exports);
var import_Tile = __toESM(require("ol/layer/Tile.js"), 1);
var import_ImageTile = __toESM(require("ol/source/ImageTile.js"), 1);
var import_data = require("@india-boundary-corrector/data");
var import_layer_configs = require("@india-boundary-corrector/layer-configs");
var import_tilefixer = require("@india-boundary-corrector/tilefixer");
var import_layer_configs2 = require("@india-boundary-corrector/layer-configs");
var import_data2 = require("@india-boundary-corrector/data");
function createCorrectedTileLoader(urlTemplate, tileFixer, layerConfig, layerRef, fetchOptions, fallbackOnCorrectionFailure) {
  return async function(z, x, y, { signal }) {
    const src = urlTemplate.replace("{z}", z).replace("{x}", x).replace("{y}", y);
    try {
      const { data, correctionsFailed, correctionsError } = await tileFixer.fetchAndFixTile(
        src,
        z,
        x,
        y,
        layerConfig,
        { signal, ...fetchOptions },
        fallbackOnCorrectionFailure
      );
      if (correctionsFailed && layerRef.current) {
        console.warn("[IndiaBoundaryCorrectedTileLayer] Corrections fetch failed:", correctionsError);
        layerRef.current.dispatchEvent({ type: "correctionerror", error: correctionsError, coords: { z, x, y }, tileUrl: src });
      }
      const blob = new Blob([data]);
      return createImageBitmap(blob);
    } catch (err) {
      if (err.name !== "AbortError") {
        console.warn("[IndiaBoundaryCorrectedTileLayer] Tile fetch failed:", err);
      }
      throw err;
    }
  };
}
var IndiaBoundaryCorrectedTileLayer = class extends import_Tile.default {
  /**
   * @param {Object} options - Layer options
   * @param {string} options.url - Tile URL template with {x}, {y}, {z} placeholders
   * @param {string} [options.pmtilesUrl] - URL to PMTiles file (defaults to CDN)
   * @param {Object|string} [options.layerConfig] - LayerConfig or config ID (auto-detected if not provided)
   * @param {Object[]} [options.extraLayerConfigs] - Additional LayerConfigs for matching
   * @param {number} [options.tileSize=256] - Tile size in pixels (for OpenLayers source)
   * @param {boolean} [options.fallbackOnCorrectionFailure=true] - Return original tile if corrections fail
   * @param {Object} [options.sourceOptions] - Additional options for ImageTile source
   * @param {string} [options.sourceOptions.crossOrigin] - Cross-origin attribute ('anonymous' or 'use-credentials')
   * @param {string} [options.sourceOptions.referrerPolicy] - Referrer policy for fetch requests
   * @param {Object} [options.layerOptions] - Additional options for TileLayer
   */
  constructor(options) {
    const {
      url,
      pmtilesUrl,
      layerConfig,
      extraLayerConfigs,
      tileSize = 256,
      fallbackOnCorrectionFailure = true,
      sourceOptions = {},
      ...layerOptions
    } = options;
    const registry = import_layer_configs.layerConfigs.createMergedRegistry(extraLayerConfigs);
    let resolvedConfig;
    if (typeof layerConfig === "string") {
      resolvedConfig = registry.get(layerConfig);
    } else if (layerConfig) {
      resolvedConfig = layerConfig;
    } else {
      resolvedConfig = registry.detectFromTemplates([url]);
    }
    const tileFixer = import_tilefixer.TileFixer.getOrCreate(pmtilesUrl ?? (0, import_data.getPmtilesUrl)());
    const crossOrigin = sourceOptions.crossOrigin ?? "anonymous";
    const fetchOptions = (0, import_tilefixer.buildFetchOptions)(crossOrigin, sourceOptions.referrerPolicy);
    const layerRef = { current: null };
    const source = new import_ImageTile.default({
      tileSize,
      crossOrigin,
      ...sourceOptions,
      loader: resolvedConfig ? createCorrectedTileLoader(url, tileFixer, resolvedConfig, layerRef, fetchOptions, fallbackOnCorrectionFailure) : void 0,
      url: resolvedConfig ? void 0 : url
    });
    super({
      source,
      ...layerOptions
    });
    layerRef.current = this;
    this._tileFixer = tileFixer;
    this._layerConfig = resolvedConfig;
    this._registry = registry;
    if (!resolvedConfig) {
      console.warn("[IndiaBoundaryCorrectedTileLayer] Could not detect layer config from URL. Corrections will not be applied.");
    }
  }
  /**
   * Get the TileFixer instance.
   * @returns {TileFixer}
   */
  getTileFixer() {
    return this._tileFixer;
  }
  /**
   * Get the resolved LayerConfig.
   * @returns {Object|null}
   */
  getLayerConfig() {
    return this._layerConfig;
  }
  /**
   * Get the registry.
   * @returns {LayerConfigRegistry}
   */
  getRegistry() {
    return this._registry;
  }
};
function indiaBoundaryCorrectedTileLayer(options) {
  return new IndiaBoundaryCorrectedTileLayer(options);
}
