// @india-boundary-corrector/layer-configs/dist/index.js
// src/configs.json
var configs_default = [
  {
    id: "cartodb-dark",
    tileUrlTemplates: [
      "https://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/dark_all/{z}/{x}/{y}.png",
      "https://{s}.basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/dark_nolabels/{z}/{x}/{y}.png"
    ],
    lineWidthStops: { "0": 1, "2": 1, "3": 1, "10": 2, "12": 2 },
    lineStyles: [
      { color: "rgb(40, 40, 40)", layerSuffix: "ne", endZoom: 4 },
      { color: "rgb(40, 40, 40)", layerSuffix: "ne-disp", endZoom: 4 },
      { color: "rgb(40, 40, 40)", layerSuffix: "osm", startZoom: 5 },
      { color: "rgb(40, 40, 40)", layerSuffix: "osm-disp", startZoom: 5 },
      { color: "rgb(40, 40, 40)", layerSuffix: "osm-internal", startZoom: 5, widthFraction: 0.3, dashArray: [2, 2], delWidthFactor: 2 },
      { color: "rgb(40, 40, 40)", layerSuffix: "ne-internal", startZoom: 4, endZoom: 4, widthFraction: 0.3, dashArray: [2, 2], delWidthFactor: 3 }
    ]
  },
  {
    id: "cartodb-dark-retina",
    tileUrlTemplates: [
      "https://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/dark_all/{z}/{x}/{y}{r}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/dark_all/{z}/{x}/{y}{r}.png",
      "https://{s}.basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}{r}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/dark_nolabels/{z}/{x}/{y}{r}.png"
    ],
    lineWidthStops: { "0": 1.5, "2": 1.5, "3": 1.5, "6": 2.5, "10": 3 },
    lineStyles: [
      { color: "rgb(40, 40, 40)", layerSuffix: "ne", endZoom: 4 },
      { color: "rgb(40, 40, 40)", layerSuffix: "ne-disp", endZoom: 4 },
      { color: "rgb(40, 40, 40)", layerSuffix: "osm", startZoom: 5 },
      { color: "rgb(40, 40, 40)", layerSuffix: "osm-disp", startZoom: 5 },
      { color: "rgb(40, 40, 40)", layerSuffix: "osm-internal", startZoom: 5, widthFraction: 0.3, dashArray: [4, 4], delWidthFactor: 2 },
      { color: "rgb(40, 40, 40)", layerSuffix: "ne-internal", startZoom: 4, endZoom: 4, widthFraction: 0.3, dashArray: [4, 4], delWidthFactor: 3 }
    ]
  },
  {
    id: "cartodb-light",
    tileUrlTemplates: [
      "https://{s}.basemaps.cartocdn.com/light_all/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/light_all/{z}/{x}/{y}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
      "https://{s}.basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_nolabels/{z}/{x}/{y}.png",
      "https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}.png",
      "https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png",
      "https://{s}.basemaps.cartocdn.com/rastertiles/voyager_labels_under/{z}/{x}/{y}.png",
      "https://basemaps.cartocdn.com/rastertiles/voyager_labels_under/{z}/{x}/{y}.png"
    ],
    lineWidthStops: { "1": 0.5, "2": 0.75, "3": 0.75, "4": 1, "5": 1, "7": 1.75, "16": 2.5 },
    lineStyles: [
      { color: "rgb(235, 214, 214)", layerSuffix: "ne", endZoom: 4, lineExtensionFactor: 0.1, delWidthFactor: 2.5 },
      { color: "rgb(235, 214, 214)", layerSuffix: "ne-disp", endZoom: 4, delWidthFactor: 0 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm", alpha: 0.1, startZoom: 6, lineWidthStops: { "6": 9, "10": 8, "14": 8 }, lineExtensionFactor: 0.1 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm", startZoom: 5, lineExtensionFactor: 0.1 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm-disp", startZoom: 5, delWidthFactor: 0 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm-internal", startZoom: 5, widthFraction: 0.5, dashArray: [2, 2], delWidthFactor: 2 },
      { color: "rgb(235, 214, 214)", layerSuffix: "ne-internal", startZoom: 4, endZoom: 4, widthFraction: 0.5, dashArray: [2, 2], delWidthFactor: 2.5 }
    ]
  },
  {
    id: "cartodb-light-retina",
    tileUrlTemplates: [
      "https://{s}.basemaps.cartocdn.com/light_all/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/light_all/{z}/{x}/{y}{r}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}{r}.png",
      "https://{s}.basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}{r}.png",
      "https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_nolabels/{z}/{x}/{y}{r}.png",
      "https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png",
      "https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}{r}.png",
      "https://{s}.basemaps.cartocdn.com/rastertiles/voyager_labels_under/{z}/{x}/{y}{r}.png",
      "https://basemaps.cartocdn.com/rastertiles/voyager_labels_under/{z}/{x}/{y}{r}.png"
    ],
    lineWidthStops: { "1": 1, "3": 1, "4": 2, "5": 2, "6": 2.5, "7": 3, "8": 3, "10": 4, "11": 4, "12": 4 },
    lineStyles: [
      { color: "rgb(235, 214, 214)", layerSuffix: "ne", endZoom: 4, lineExtensionFactor: 0.1, delWidthFactor: 2.5 },
      { color: "rgb(235, 214, 214)", layerSuffix: "ne-disp", endZoom: 4, delWidthFactor: 0 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm", alpha: 0.1, startZoom: 6, lineWidthStops: { "6": 20, "8": 19, "10": 18, "12": 17 }, lineExtensionFactor: 0.1 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm", startZoom: 5, lineExtensionFactor: 0.1 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm-disp", startZoom: 5, delWidthFactor: 0 },
      { color: "rgb(235, 214, 214)", layerSuffix: "osm-internal", startZoom: 5, widthFraction: 0.5, dashArray: [4, 4], delWidthFactor: 2 },
      { color: "rgb(235, 214, 214)", layerSuffix: "ne-internal", startZoom: 4, endZoom: 4, widthFraction: 0.5, dashArray: [4, 4], delWidthFactor: 2.5 }
    ]
  },
  {
    id: "open-topo",
    tileUrlTemplates: [
      "https://{s}.tile.opentopomap.org/{z}/{x}/{y}.png",
      "https://tile.opentopomap.org/{z}/{x}/{y}.png",
      "https://{s}.tile.top-o-map.de/{z}/{x}/{y}.png",
      "https://tile.top-o-map.de/{z}/{x}/{y}.png"
    ],
    lineWidthStops: { "4": 0.75, "5": 1, "6": 1.25, "7": 1.5, "8": 1.75, "9": 1.25, "10": 1.25, "13": 1.5 },
    lineStyles: [
      { color: "rgb(83, 83, 83)", layerSuffix: "ne", startZoom: 4, endZoom: 6 },
      { color: "rgb(173, 173, 173)", layerSuffix: "osm", startZoom: 7, endZoom: 8, alpha: 0.5, widthFraction: 4 },
      { color: "rgb(83, 83, 83)", layerSuffix: "osm", startZoom: 7, endZoom: 8 },
      { color: "rgb(83, 83, 83)", layerSuffix: "osm-internal", widthFraction: 0.75, startZoom: 7, endZoom: 8 },
      { color: "rgb(199, 158, 204)", layerSuffix: "osm", startZoom: 9, widthFraction: 7, alpha: 0.6, lineExtensionFactor: 0.2 },
      { color: "rgb(175, 41, 203)", layerSuffix: "osm", startZoom: 9, lineExtensionFactor: 0.2 },
      { color: "rgb(199, 158, 204)", layerSuffix: "osm-internal", startZoom: 9, widthFraction: 2.25, alpha: 0.6, lineExtensionFactor: 0.2 },
      { color: "rgb(175, 41, 203)", layerSuffix: "osm-internal", startZoom: 9, widthFactor: 0.75, lineExtensionFactor: 0.2 }
    ]
  },
  {
    id: "osm-carto",
    tileUrlTemplates: [
      "https://tile.openstreetmap.org/{z}/{x}/{y}.png",
      "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
    ],
    lineWidthStops: { "3": 0.7, "4": 1, "10": 3.75 },
    lineStyles: [
      { color: "rgb(200, 180, 200)", layerSuffix: "osm", startZoom: 4, delWidthFactor: 1.5 },
      { color: "rgb(160, 120, 160)", layerSuffix: "osm", startZoom: 4, widthFraction: 0.333, dashArray: [30, 2, 8, 2], delWidthFactor: 0 },
      { color: "rgb(200, 180, 200)", layerSuffix: "osm-internal", startZoom: 4, widthFraction: 0.45 },
      { color: "rgb(160, 120, 160)", layerSuffix: "osm-internal", startZoom: 4, widthFraction: 0.15, dashArray: [10, 4, 2] }
    ]
  },
  {
    id: "osm-hot",
    tileUrlTemplates: [
      "https://{s}.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
      "https://tile.openstreetmap.fr/hot/{z}/{x}/{y}.png"
    ],
    lineWidthStops: { "2": 3, "3": 3, "7": 3, "8": 3.5, "9": 3.5 },
    lineStyles: [
      { color: "rgb(149, 175, 180)", layerSuffix: "osm", lineExtensionFactor: 0.25, startZoom: 2 },
      { color: "rgb(89, 117, 123)", layerSuffix: "osm", widthFraction: 0.33, lineExtensionFactor: 0.25, startZoom: 2 },
      { color: "rgb(172, 163, 163)", layerSuffix: "osm-internal", widthFraction: 0.33, startZoom: 4, dashArray: [7, 7] }
    ]
  }
];

// src/layerconfig.js
var INFINITY = -1;
var MIN_LINE_WIDTH = 0.1;
var DEFAULT_LINE_WIDTH = 1;
function interpolateLineWidth(zoom, lineWidthStops) {
  const zooms = Object.keys(lineWidthStops).map(Number).sort((a, b) => a - b);
  if (lineWidthStops[zoom] !== void 0) {
    return lineWidthStops[zoom];
  }
  if (zoom < zooms[0]) {
    const z1 = zooms[0];
    const z2 = zooms[1];
    const w1 = lineWidthStops[z1];
    const w2 = lineWidthStops[z2];
    const slope = (w2 - w1) / (z2 - z1);
    return Math.max(MIN_LINE_WIDTH, w1 + slope * (zoom - z1));
  }
  if (zoom > zooms[zooms.length - 1]) {
    const z1 = zooms[zooms.length - 2];
    const z2 = zooms[zooms.length - 1];
    const w1 = lineWidthStops[z1];
    const w2 = lineWidthStops[z2];
    const slope = (w2 - w1) / (z2 - z1);
    return Math.max(MIN_LINE_WIDTH, w2 + slope * (zoom - z2));
  }
  for (let i = 0; i < zooms.length - 1; i++) {
    if (zoom > zooms[i] && zoom < zooms[i + 1]) {
      const z1 = zooms[i];
      const z2 = zooms[i + 1];
      const w1 = lineWidthStops[z1];
      const w2 = lineWidthStops[z2];
      const t = (zoom - z1) / (z2 - z1);
      return w1 + t * (w2 - w1);
    }
  }
  return DEFAULT_LINE_WIDTH;
}
function templateToRegex(template) {
  const groups = [];
  const hasRetina = template.includes("{r}");
  const hasExtension = /\.(png|jpg|jpeg|webp|gif)$/i.test(template);
  let pattern = template.replace(/[.*+?^${}()|[\]\\]/g, (char) => {
    if (char === "{" || char === "}") return char;
    return "\\" + char;
  }).replace(/^https:\/\//, "https?://").replace(/^http:\/\//, "https?://").replace(/\{[a-z0-9]-[a-z0-9]\}/gi, () => {
    groups.push("s");
    return "([a-z0-9]+)";
  }).replace(/\{(z|x|y|s|r)\}/gi, (_, name) => {
    const lowerName = name.toLowerCase();
    groups.push(lowerName);
    if (lowerName === "s") {
      return "([a-z0-9]+)";
    }
    if (lowerName === "r") {
      return "(@\\d+x)";
    }
    return "(\\d+)";
  });
  if (!hasRetina) {
    if (hasExtension) {
      pattern = pattern.replace(/(\\\.(png|jpg|jpeg|webp|gif))/, "(?!@\\d+x)$1");
    } else {
      pattern = pattern + "(?!@\\d+x)";
    }
  }
  return { pattern: new RegExp("^" + pattern + "(\\?.*)?$", "i"), groups };
}
function templateToTemplateRegex(template) {
  const hasRetina = template.includes("{r}");
  const hasExtension = /\.(png|jpg|jpeg|webp|gif)$/i.test(template);
  let pattern = template.replace(/[.*+?^${}()|[\]\\]/g, (char) => {
    if (char === "{" || char === "}") return char;
    return "\\" + char;
  }).replace(/^https:\/\//, "https?://").replace(/^http:\/\//, "https?://").replace(/\{([a-z0-9])-([a-z0-9])\}/gi, (_, start, end) => `(\\{${start}-${end}\\}|\\{s\\}|[a-z0-9]+)`).replace(/\{(z|x|y|s|r)\}/gi, (_, name) => {
    const lowerName = name.toLowerCase();
    if (lowerName === "s") {
      return "(\\{s\\}|\\{[a-z0-9]-[a-z0-9]\\}|[a-z0-9]+)";
    }
    if (lowerName === "r") {
      return "(\\{r\\}|@\\d+x)";
    }
    return `\\{${lowerName}\\}`;
  });
  if (!hasRetina) {
    if (hasExtension) {
      pattern = pattern.replace(/(\\\.(png|jpg|jpeg|webp|gif))/, "(?!@\\d+x|\\{r\\})$1");
    } else {
      pattern = pattern + "(?!@\\d+x|\\{r\\})";
    }
  }
  return new RegExp("^" + pattern + "(\\?.*)?$", "i");
}
function isValidColor(color) {
  if (typeof color !== "string" || !color.trim()) return false;
  if (typeof CSS !== "undefined" && CSS.supports) {
    return CSS.supports("color", color);
  }
  const trimmed = color.trim().toLowerCase();
  if (/^#([0-9a-f]{3}|[0-9a-f]{6}|[0-9a-f]{8})$/.test(trimmed)) return true;
  if (/^(rgb|hsl)a?\(/.test(trimmed)) return true;
  if (/^[a-z]+$/.test(trimmed)) return true;
  return false;
}
function validateLineWidthStops(lineWidthStops, prefix) {
  if (!lineWidthStops || typeof lineWidthStops !== "object" || Array.isArray(lineWidthStops)) {
    throw new Error(`${prefix}: lineWidthStops must be an object`);
  }
  const stopKeys = Object.keys(lineWidthStops);
  if (stopKeys.length < 2) {
    throw new Error(`${prefix}: lineWidthStops must have at least 2 entries`);
  }
  for (const key of stopKeys) {
    const zoom = Number(key);
    if (!Number.isInteger(zoom) || zoom < 0) {
      throw new Error(`${prefix}: lineWidthStops keys must be non-negative integers, got "${key}"`);
    }
    if (typeof lineWidthStops[key] !== "number" || lineWidthStops[key] <= 0) {
      throw new Error(`${prefix}: lineWidthStops values must be positive numbers`);
    }
  }
}
var LineStyle = class _LineStyle {
  /**
   * Validate a LineStyle configuration object.
   * @param {Object} obj - The object to validate
   * @param {number} [index] - Optional index for error messages (when validating in an array)
   * @param {boolean} [requireLineWidthStops=false] - Whether lineWidthStops is required
   * @throws {Error} If validation fails
   */
  static validateJSON(obj, index, requireLineWidthStops = false) {
    const prefix = index !== void 0 ? `lineStyles[${index}]` : "LineStyle";
    if (!obj || typeof obj !== "object") {
      throw new Error(`${prefix}: must be an object`);
    }
    if (!obj.color || typeof obj.color !== "string") {
      throw new Error(`${prefix}: color must be a non-empty string`);
    }
    if (!isValidColor(obj.color)) {
      throw new Error(`${prefix}: color "${obj.color}" is not a valid CSS color`);
    }
    if (!obj.layerSuffix || typeof obj.layerSuffix !== "string") {
      throw new Error(`${prefix}: layerSuffix must be a non-empty string`);
    }
    if (obj.widthFraction !== void 0 && (typeof obj.widthFraction !== "number" || obj.widthFraction <= 0)) {
      throw new Error(`${prefix}: widthFraction must be a positive number`);
    }
    if (obj.dashArray !== void 0 && !Array.isArray(obj.dashArray)) {
      throw new Error(`${prefix}: dashArray must be an array`);
    }
    if (obj.alpha !== void 0 && (typeof obj.alpha !== "number" || obj.alpha < 0 || obj.alpha > 1)) {
      throw new Error(`${prefix}: alpha must be a number between 0 and 1`);
    }
    if (obj.startZoom !== void 0 && (typeof obj.startZoom !== "number" || obj.startZoom < 0)) {
      throw new Error(`${prefix}: startZoom must be a non-negative number`);
    }
    if (obj.endZoom !== void 0 && (typeof obj.endZoom !== "number" || obj.endZoom < 0 && obj.endZoom !== INFINITY)) {
      throw new Error(`${prefix}: endZoom must be a non-negative number or INFINITY (${INFINITY})`);
    }
    if (obj.lineExtensionFactor !== void 0 && (typeof obj.lineExtensionFactor !== "number" || obj.lineExtensionFactor < 0)) {
      throw new Error(`${prefix}: lineExtensionFactor must be a non-negative number`);
    }
    if (obj.delWidthFactor !== void 0 && (typeof obj.delWidthFactor !== "number" || obj.delWidthFactor < 0)) {
      throw new Error(`${prefix}: delWidthFactor must be a non-negative number`);
    }
    if (requireLineWidthStops && obj.lineWidthStops === void 0) {
      throw new Error(`${prefix}: lineWidthStops is required`);
    }
    if (obj.lineWidthStops !== void 0) {
      validateLineWidthStops(obj.lineWidthStops, prefix);
    }
  }
  /**
   * @param {Object} options
   * @param {string} options.color - CSS color string
   * @param {string} options.layerSuffix - Layer suffix (e.g., 'osm', 'ne', 'osm-disp')
   * @param {Object<number, number>} options.lineWidthStops - Line width stops for this style
   * @param {number} [options.widthFraction=1.0] - Multiplier for base line width
   * @param {number[]} [options.dashArray] - Dash pattern for dashed lines
   * @param {number} [options.alpha=1.0] - Opacity (0-1)
   * @param {number} [options.startZoom=0] - Minimum zoom level for this style
   * @param {number} [options.endZoom=INFINITY] - Maximum zoom level for this style (INFINITY means no limit)
   * @param {number} [options.lineExtensionFactor=0.0] - Factor to extend lines by (multiplied by deletion line width)
   * @param {number} [options.delWidthFactor=1.5] - Factor to multiply line width for deletion blur
   */
  constructor({ color, layerSuffix, lineWidthStops, widthFraction = 1, dashArray, alpha = 1, startZoom = 0, endZoom = INFINITY, lineExtensionFactor = 0, delWidthFactor = 1.5 }) {
    this.color = color;
    this.layerSuffix = layerSuffix;
    this.lineWidthStops = lineWidthStops;
    this.widthFraction = widthFraction;
    this.dashArray = dashArray;
    this.alpha = alpha;
    this.startZoom = startZoom;
    this.endZoom = endZoom;
    this.lineExtensionFactor = lineExtensionFactor;
    this.delWidthFactor = delWidthFactor;
  }
  /**
   * Get base line width for this style at a given zoom level.
   * @param {number} zoom - Zoom level
   * @returns {number}
   */
  getLineWidth(zoom) {
    return interpolateLineWidth(zoom, this.lineWidthStops);
  }
  /**
   * Check if this style is active at the given zoom level.
   * @param {number} z - Zoom level
   * @returns {boolean}
   */
  isActiveAtZoom(z) {
    return z >= this.startZoom && (this.endZoom === INFINITY || z <= this.endZoom);
  }
  /**
   * Serialize to plain object.
   * @returns {Object}
   */
  toJSON() {
    return {
      color: this.color,
      layerSuffix: this.layerSuffix,
      lineWidthStops: this.lineWidthStops,
      widthFraction: this.widthFraction,
      dashArray: this.dashArray,
      alpha: this.alpha,
      startZoom: this.startZoom,
      endZoom: this.endZoom,
      lineExtensionFactor: this.lineExtensionFactor,
      delWidthFactor: this.delWidthFactor
    };
  }
  /**
   * Create from plain object with validation.
   * @param {Object} obj
   * @param {number} [index] - Optional index for error messages
   * @returns {LineStyle}
   */
  static fromJSON(obj, index) {
    _LineStyle.validateJSON(obj, index, true);
    return new _LineStyle(obj);
  }
};
var LayerConfig = class _LayerConfig {
  /**
   * Validate a LayerConfig configuration object.
   * Also validates all lineStyles within the config.
   * @param {Object} obj - The object to validate
   * @throws {Error} If validation fails
   */
  static validateJSON(obj) {
    if (!obj || typeof obj !== "object") {
      throw new Error("LayerConfig: must be an object");
    }
    if (!obj.id || typeof obj.id !== "string") {
      throw new Error("LayerConfig: id must be a non-empty string");
    }
    if (obj.id.includes("/")) {
      throw new Error(`LayerConfig: id cannot contain slashes: "${obj.id}"`);
    }
    const id = obj.id;
    if (obj.lineWidthStops !== void 0) {
      if (!obj.lineWidthStops || typeof obj.lineWidthStops !== "object" || Array.isArray(obj.lineWidthStops)) {
        throw new Error(`LayerConfig "${id}": lineWidthStops must be an object`);
      }
      const stopKeys = Object.keys(obj.lineWidthStops);
      if (stopKeys.length < 2) {
        throw new Error(`LayerConfig "${id}": lineWidthStops must have at least 2 entries`);
      }
      for (const key of stopKeys) {
        const zoom = Number(key);
        if (!Number.isInteger(zoom) || zoom < 0) {
          throw new Error(`LayerConfig "${id}": lineWidthStops keys must be non-negative integers, got "${key}"`);
        }
        if (typeof obj.lineWidthStops[key] !== "number" || obj.lineWidthStops[key] <= 0) {
          throw new Error(`LayerConfig "${id}": lineWidthStops values must be positive numbers`);
        }
      }
    }
    if (!Array.isArray(obj.lineStyles) || obj.lineStyles.length === 0) {
      throw new Error(`LayerConfig "${id}": lineStyles must be a non-empty array`);
    }
    for (let i = 0; i < obj.lineStyles.length; i++) {
      LineStyle.validateJSON(obj.lineStyles[i], i);
    }
  }
  constructor({
    id,
    // Tile URL templates for matching (e.g., "https://{s}.tile.example.com/{z}/{x}/{y}.png")
    tileUrlTemplates = [],
    // Line width stops: map of zoom level to line width (at least 2 entries)
    // Note: interpolated/extrapolated line width is capped at a minimum of 0.5
    lineWidthStops = { 1: 0.5, 10: 2.5 },
    // Line styles array - each element describes a line to draw from a specific layer
    // { color: string, layerSuffix: string, widthFraction?: number, dashArray?: number[], 
    //   startZoom?: number, endZoom?: number, lineExtensionFactor?: number, delWidthFactor?: number }
    // Lines are drawn in array order. layerSuffix determines which PMTiles layer to use.
    lineStyles
  }) {
    this.id = id;
    const templates = Array.isArray(tileUrlTemplates) ? tileUrlTemplates : tileUrlTemplates ? [tileUrlTemplates] : [];
    this.tileUrlTemplates = templates;
    this._compiledPatterns = templates.map((t) => templateToRegex(t));
    this._templatePatterns = templates.map((t) => templateToTemplateRegex(t));
    this.lineWidthStops = lineWidthStops;
    this.lineStyles = lineStyles.map((style) => {
      if (style instanceof LineStyle) {
        return style;
      }
      const styleWithStops = style.lineWidthStops ? style : { ...style, lineWidthStops };
      return new LineStyle(styleWithStops);
    });
  }
  /**
   * Get line styles active at a given zoom level
   * @param {number} z - Zoom level
   * @returns {LineStyle[]}
   */
  getLineStylesForZoom(z) {
    return this.lineStyles.filter((style) => style.isActiveAtZoom(z));
  }
  /**
   * Get unique layer suffixes from styles active at a given zoom level
   * @param {number} z - Zoom level
   * @returns {string[]}
   */
  getLayerSuffixesForZoom(z) {
    const activeStyles = this.getLineStylesForZoom(z);
    return [...new Set(activeStyles.map((s) => s.layerSuffix))];
  }
  /**
   * Check if this config matches the given template URLs (with {z}/{x}/{y} placeholders)
   * @param {string | string[]} templates - Single template URL or array of template URLs
   * @returns {boolean}
   */
  matchTemplate(templates) {
    if (this._templatePatterns.length === 0) return false;
    const urls = Array.isArray(templates) ? templates : [templates];
    if (urls.length === 0) return false;
    return urls.some(
      (url) => this._templatePatterns.some((pattern) => pattern.test(url))
    );
  }
  /**
   * Check if this config matches the given tile URLs (with actual coordinates)
   * @param {string | string[]} tiles - Single tile URL or array of tile URLs
   * @returns {boolean}
   */
  matchTileUrl(tiles) {
    if (this._compiledPatterns.length === 0) return false;
    const urls = Array.isArray(tiles) ? tiles : [tiles];
    if (urls.length === 0) return false;
    return urls.some(
      (url) => this._compiledPatterns.some(({ pattern }) => pattern.test(url))
    );
  }
  /**
   * Extract tile coordinates (z, x, y) from a URL using this config's templates
   * @param {string} url - Tile URL to extract coordinates from
   * @returns {{ z: number, x: number, y: number } | null}
   */
  extractCoords(url) {
    for (const { pattern, groups } of this._compiledPatterns) {
      const match = url.match(pattern);
      if (match) {
        const result = {};
        for (let i = 0; i < groups.length; i++) {
          const name = groups[i];
          const value = match[i + 1];
          if (name === "z" || name === "x" || name === "y") {
            result[name] = parseInt(value, 10);
          }
        }
        if ("z" in result && "x" in result && "y" in result) {
          return { z: result.z, x: result.x, y: result.y };
        }
      }
    }
    return null;
  }
  /**
   * Serialize the config to a plain object for postMessage
   * @returns {Object}
   */
  toJSON() {
    return {
      id: this.id,
      tileUrlTemplates: this.tileUrlTemplates,
      lineWidthStops: this.lineWidthStops,
      lineStyles: this.lineStyles.map((s) => s.toJSON())
    };
  }
  /**
   * Create a LayerConfig from a plain object with validation.
   * @param {Object} obj
   * @returns {LayerConfig}
   * @throws {Error} If validation fails
   */
  static fromJSON(obj) {
    _LayerConfig.validateJSON(obj);
    return new _LayerConfig(obj);
  }
};

// src/index.js
var LayerConfigRegistry = class _LayerConfigRegistry {
  constructor() {
    this.registry = {};
  }
  /**
   * Get a layer config by id
   */
  get(id) {
    return this.registry[id];
  }
  /**
   * Register a new layer config
   */
  register(config) {
    this.registry[config.id] = config;
  }
  /**
   * Remove a layer config by id
   */
  remove(id) {
    if (!this.registry[id]) return false;
    delete this.registry[id];
    return true;
  }
  /**
   * Detect layer config from tile URL templates (with {z}/{x}/{y} placeholders)
   * @param {string | string[]} templates - Single template URL or array of template URLs
   */
  detectFromTemplates(templates) {
    if (!templates || Array.isArray(templates) && templates.length === 0) return void 0;
    for (const config of Object.values(this.registry)) {
      if (config.matchTemplate(templates)) {
        return config;
      }
    }
    return void 0;
  }
  /**
   * Detect layer config from actual tile URLs (with numeric coordinates)
   * @param {string | string[]} urls - Single tile URL or array of tile URLs
   */
  detectFromTileUrls(urls) {
    if (!urls || Array.isArray(urls) && urls.length === 0) return void 0;
    for (const config of Object.values(this.registry)) {
      if (config.matchTileUrl(urls)) {
        return config;
      }
    }
    return void 0;
  }
  /**
   * Get all available layer config ids
   */
  getAvailableIds() {
    return Object.keys(this.registry);
  }
  /**
   * Create a new registry with all configs from this registry plus extra configs.
   * @param {LayerConfig[]} extraLayerConfigs - Additional configs to add
   * @returns {LayerConfigRegistry} A new registry with merged configs
   */
  createMergedRegistry(extraLayerConfigs) {
    const registry = new _LayerConfigRegistry();
    for (const id of this.getAvailableIds()) {
      registry.register(this.get(id));
    }
    if (extraLayerConfigs && extraLayerConfigs.length > 0) {
      for (const config of extraLayerConfigs) {
        registry.register(config);
      }
    }
    return registry;
  }
  /**
   * Parse a tile URL into its components: layer config and coordinates
   * @param {string} url - Tile URL to parse
   * @returns {{ layerConfig: LayerConfig, coords: { z: number, x: number, y: number } } | null}
   */
  parseTileUrl(url) {
    const layerConfig = this.detectFromTileUrls([url]);
    if (!layerConfig) return null;
    const coords = layerConfig.extractCoords(url);
    if (!coords) return null;
    return { layerConfig, coords };
  }
};
var layerConfigs = new LayerConfigRegistry();
for (const configData of configs_default) {
  layerConfigs.register(new LayerConfig(configData));
}
export {
  INFINITY,
  LayerConfig,
  LayerConfigRegistry,
  LineStyle,
  MIN_LINE_WIDTH,
  configs_default as configsJson,
  interpolateLineWidth,
  layerConfigs
};
